### PR TITLE
[@mantine/hooks] use-state-validator: init hook

### PR DESCRIPTION
**Pull request by achmurali**
This PR implements the use-state-validator hook.

yet to add documentation, shall add it once reviewed

### CODE REVIEW

File: src/mantine-hooks/src/use-state-validator/use-state-validator.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import { useState } from 'react';
+
+export function useStateValidator<T>(
+  initialValue: T | null | undefined,
+  validator: (val: T | null | undefined) => boolean
+) {
+  let isValid = validator(initialValue);
+  const [valid, setValid] = useState<boolean>(isValid);
+  const [value, setValue] = useState<T | null | undefined>(isValid ? initialValue : undefined);
```
Comment (rtivital) [inline src/mantine-hooks/src/use-state-validator/use-state-validator.ts:9]: initialValue should always be set, no matter of validation state, same in handleChange function

File: src/mantine-hooks/src/use-state-validator/use-state-validator.test.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
+/* eslint-disable no-console */
+import { act, renderHook } from '@testing-library/react-hooks';
+import { useStateValidator } from './use-state-validator';
+
+describe('@mantine/hooks/use-state-validator', () => {
+  beforeEach(() => {
+    jest.clearAllMocks();
```
Comment (rtivital) [inline src/mantine-hooks/src/use-state-validator/use-state-validator.test.ts:7]: Why do we need to clear mocks?

File: src/mantine-hooks/src/use-state-validator/use-state-validator.test.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
+/* eslint-disable no-console */
```
Comment (rtivital) [inline src/mantine-hooks/src/use-state-validator/use-state-validator.test.ts:1]: Redundant no-console

File: src/mantine-hooks/src/use-state-validator/use-state-validator.ts (hunk 1 above)
Comment (achmurali) [inline src/mantine-hooks/src/use-state-validator/use-state-validator.ts:9]: I have used `useState` for managing state, should I change it to useRef and push updates manually using useForceUpdate and etc?

Comment (rtivital) [reply src/mantine-hooks/src/use-state-validator/use-state-validator.ts:9]: No, it should be use-state

**Comment by jerebtw** [comment]
I have added such a hook in another pr https://github.com/mantinedev/mantine/pull/586

https://github.com/mantinedev/mantine/pull/586/commits/f995808a1246a47707f2be26d3b063cb06229ae0

**Comment by rtivital** [comment]
@jerebtw please discuss in Discord before adding anything new, the hook that you've added is not the same and works differently, I would prefer the implementation of @achmurali because we've intended to add this for some time

**Comment by jerebtw** [comment]
Ohh okay, I needed that feature for the am/pm so added a hook. And I think we can add both because my hook is quit different from this.

**Pull request closed without merging**